lib/elixir/lib/keyword.ex

```diff
@@ -69,9 +69,7 @@ defmodule Keyword do
   """
   @spec new(Enum.t) :: t
   def new(pairs) do
-    Enum.reduce pairs, [], fn {k, v}, keywords ->
-      put(keywords, k, v)
-    end
+    Enum.uniq_by(Enum.reverse(pairs), fn {x, _} when is_atom(x) -> x end)
```

Comment:
@josevalim I've done sanity check for this.
As expected, `HasSet`-based implementation's starting to be more performant; in my test after ~1700 elements (e.g. for 10k elements: `list-based: 72078 µs` and `HasSet-based: 13504 µs`).